mbedtls/src/x509/certificate.rs

```diff
@@ -935,28 +1051,464 @@ cYp0bH/RcPTC0Z+ZaqSWMtfxRrk63MJQF9EXpDCdvQRcTMD9D85DJrMKn8aumq0M
+            let it_clone = chain.clone();
+            let mut it = it_clone.iter();
+
+            assert_eq!(c1_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c2_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c3_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c1_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c2_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c3_info, format!("{:?}", it.next().unwrap()));
+
+            assert!(it.next().is_none());
+            assert!(it.next().is_none());
+            assert!(it.next().is_none());
         }
     }
+
+    #[test]
+    fn list_into_iter() {
+        const C_LEAF: &'static str = concat!(include_str!("../../tests/data/chain-leaf.crt"),"\0");
+        const C_INT1: &'static str = concat!(include_str!("../../tests/data/chain-int1.crt"),"\0");
+        const C_INT2: &'static str = concat!(include_str!("../../tests/data/chain-int2.crt"),"\0");
+
+        let mut chain = MbedtlsList::<Certificate>::new();
+
+        let c1 = Certificate::from_pem_multiple(C_LEAF.as_bytes()).unwrap();
+        let c2 = Certificate::from_pem_multiple(C_INT1.as_bytes()).unwrap();
+        let c3 = Certificate::from_pem_multiple(C_INT2.as_bytes()).unwrap();
+
+        // Using debug strings so failing unit tests can identify any issue with contents.
+        let c1_info = format!("{:?}", c1.iter().next().unwrap());
+        let c2_info = format!("{:?}", c2.iter().next().unwrap());
+        let c3_info = format!("{:?}", c3.iter().next().unwrap());
+
+        println!("c1: {:?}", c1_info);
+        println!("c2: {:?}", c2_info);
+        println!("c3: {:?}", c3_info);
+        
+        chain.append(c1.clone());
+        chain.append(c2.clone());
+        chain.append(c3.clone());
+
+        {
+            let mut it = chain.clone().into_iter();
+            assert_eq!(c1_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c2_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c3_info, format!("{:?}", it.next().unwrap()));
+            assert!(it.next().is_none());
+        }
+
+        let clone = chain.clone();
+        
+        {
+            let mut it = chain.into_iter();
+            assert_eq!(c1_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c2_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c3_info, format!("{:?}", it.next().unwrap()));
+            assert!(it.next().is_none());
+        }
+
+        {
+            let mut it = clone.into_iter();
+            assert_eq!(c1_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c2_info, format!("{:?}", it.next().unwrap()));
+            assert_eq!(c3_info, format!("{:?}", it.next().unwrap()));
+            assert!(it.next().is_none());
+        }
+    }
+
+
+    #[test]
+    fn empty_certificate_test() {
```

Comment:
This test is not useful. All `Certificate`s should contain an actual certificate.

Reply:
This is useful to verify we have null pointer checks everywhere needed and don't crash, otherwise I would need to find a der/pem that translates to an empty but parseable certificate.

Otherwise I agree, we cannot have a certificate that was never parsed.

Reply:
re-open if you want me to remove the tests and I will